orttraining/orttraining/core/graph/gradient_builder_base.cc

```diff
@@ -165,5 +192,27 @@ void GradientBuilderBase::HandleBroadcasting(const ArgDef& input_grad,
   }
 }
 
+void GradientBuilderBase::HandleBroadcastingDynamic(const ArgDef& input_grad,
+                                                    const ArgDef& target,
+                                                    const ArgDef& output_grad,
+                                                    const ArgDef& reduce_axes,
+                                                    std::vector<NodeDef>& output) const {
+  ArgDef reduce_grad_arg = IA("ReduceSumTraining_" + input_grad.name + "_for_" + target.name);
+  output.push_back(
+      NodeDef(OpDef{"ReduceSumTraining", kMSDomain, 1},
+              {input_grad,
+               reduce_axes},
+              {reduce_grad_arg},
+              {{"keepdims", ONNX_NAMESPACE::MakeAttribute("keepdims", int64_t(1))},
+               {"noop_with_empty_axes", ONNX_NAMESPACE::MakeAttribute("noop_with_empty_axes", int64_t(1))}}));
+
+  //reuse shape node from BroadcastGradientArgs input
+  ArgDef target_shape_arg = ArgDef("Shape_" + target.name);
```

Comment:
Let's pass the ArgDef from function input, since they are not too far away. 

It's better to have explicit name passing then implicit name matching.